src/service/builder.js

```diff
@@ -0,0 +1,243 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {LayoutPriority} from '../layout';
+import {READY_SCAN_SIGNAL} from './resources-interface';
+import {VisibilityState} from '../visibility-state';
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {hasNextNodeInDocumentOrder, isIframed} from '../dom';
+import {removeItem} from '../utils/array';
+
+const ID = 'builder';
+
+/** @implements {../service.Disposable} */
+export class Builder {
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc  */
+  constructor(ampdoc) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.observer_ = new win.IntersectionObserver((e) => this.observed_(e), {
+      // Root bounds are not important, so we can use the `root:null` for a
+      // top-level window.
+      root: isIframed(win) ? win.document : null,
+      rootMargin: '250% 31.25%',
+      threshold: 0.001,
+    });
+
+    /** @private @const {!Map<!AmpElement, {asap: boolean, isIntersecting: boolean}>} */
+    this.targets_ = new Map();
+
+    /** @private {?Array<!AmpElement>} */
+    this.parsingTargets_ = [];
+
+    /** @private {boolean} */
+    ampdoc.whenReady().then(() => this.checkParsing_());
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() =>
+      this.docVisibilityChanged_()
+    );
+  }
+
+  /** @override */
+  dispose() {
+    this.observer_.disconnect();
+    this.targets_.clear();
+    if (this.visibilityUnlisten_) {
+      this.visibilityUnlisten_();
+      this.visibilityUnlisten_ = null;
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  scheduleAsap(target) {
+    this.targets_.set(target, {asap: true, isIntersecting: false});
+
+    if (target.mutable() && this.parsingTargets_) {
+      // Don't need to wait until parsing is complete.
+      removeItem(this.parsingTargets_, target);
+      this.maybeBuild_(target);
+    } else {
+      this.waitParsing_(target);
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  schedule(target) {
+    if (this.targets_.has(target)) {
+      return;
+    }
+
+    if (target.deferredBuild()) {
+      this.targets_.set(target, {asap: false, isIntersecting: false});
+      this.observer_.observe(target);
+    } else {
+      this.targets_.set(target, {asap: false, isIntersecting: true});
```

Comment:
isn't this backwards? If building cannot be deferred, then should be `asap: true`.

Reply:
There are three categories:

1. Must wait for parse and intersection
2. Must wait only for parse, but follow the priority
3. Must wait for parse, but build as soon as it's parsed.

The `asap` helps with the manually requested build where the caller needs the element "now". It only respects parsing wait, but ignores the element priority, visibility, etc.

Reply:
Can we combine (1) and (2)? Are there cases where a component cannot defer build and yet should also be scheduled via low priority? 

If you think this separation is helpful, then feel free to leave as-is, but `isIntersecting` is no longer a good word. `blockedByVisibility` is a bit closer

Reply:
I don't think we can combine (1) and (2) are sort of already combined. They both wait for parse and respect the element's priority. The only diff: (1) waits for intersection, and (2) does not.

(3) is the distinct case - someone has requested the direct build and we want to satisfy it ASAP (hence the name). We still cannot allow the element to build before it's fully parsed, but we don't need to wait for intersection.  However, I also don't think we should consult the build priority in this case. IMHO if someone asks for build, we should just build as quickly as we can. Does it make sense?